File: src/leetcode/pure_program/IsPalindrome.java
package leetcode.pure_program;

/**
 * @author chengzw
 * @description 验证回文串，https://leetcode-cn.com/problems/valid-palindrome/
 * 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写。
 * 输入: "A man, a plan, a canal: Panama"
 * 输出: true
 * 解释："amanaplanacanalpanama" 是回文串
 * @since 2021/8/17
 */
public class IsPalindrome {

    public static boolean solution1(String s) {
        char[] newChar = s.toCharArray();
        System.out.println(newChar);
        //两个指针
        int i = 0;
        int j = s.length() - 1;
        System.out.println(j);

        while (j > i) {
            // 不是数字或字⺟的话，i就⼀直++
            if (!isAlpha(newChar[i])) {
                i++;
                continue;
            }

            // 不是数字或字⺟的话，j就⼀直--
            if (!isAlpha(newChar[j])) {
                j--;
                continue;
            }

            //⾛到这⾥的话，i、j都指向数字或字⺟，看看两个字符是否相同
            if (toLower(newChar[i]) != toLower(newChar[j])) {
                return false;
            } else {
                // i和j往中间挪⼀位
                i++;
                j--;
            }
        }
        return true;
    }

    //判断是否是字母
    public static boolean isAlpha(char c) {
        if (c >= '0' && c <= '9') return true;
        if (c >= 'a' && c <= 'z') return true;
        if (c >= 'A' && c <= 'Z') return true;
        return false;
    }

    //大写转小写
    public static char toLower(char c) {
        if (c >= '0' && c <= '9') return c;
        if (c >= 'a' && c <= 'z') return c;
        //⼤写A~Z 65~90，⼩写a~z 97~122
        return (char) ((int) c + 32);
    }

    public static void main(String[] args) {
        System.out.println(solution1("A man, a plan, a canal: Panama"));
        System.out.println(solution1(".,"));
    }
}

File: go/leetcode/two_pointers/is_palindrome.go
package main

/**
 * @description 9. 回文数
 * @author chengzw
 * @since 2022/5/27
 * @link https://leetcode.cn/problems/palindrome-number/
 */

/**
给你一个整数 x ，如果 x 是一个回文整数，返回 true ；否则，返回 false 。
回文数是指正序（从左向右）和倒序（从右向左）读都是一样的整数。
例如，121 是回文，而 123 不是。

示例 1：
输入：x = 121
输出：true

示例 2：
输入：x = -121
输出：false
解释：从左向右读, 为 -121 。 从右向左读, 为 121- 。因此它不是一个回文数。

示例 3：
输入：x = 10
输出：false
解释：从右向左读, 为 01 。因此它不是一个回文数。
*/

/**
思路：
1.反转后半部分的数字，然后和前半部分进行比较。
例如，输入 1221，我们可以将数字 “1221” 的后半部分从 “21” 反转为 “12”，并将其与前半部分 “12” 进行比较，因为二者相同，我们得知数字 1221 是回文。
2.负数或者各位数为 0 时，不是回文数。
3.当原始数字小于或等于反转后的数字时，表示已经反转了一半的数字

时间复杂度：O(log n)，对于每次迭代，我们会将输入除以 10，因此时间复杂度为 O(logn)。
空间复杂度：O(1)。我们只需要常数空间存放若干变量。
*/
func isPalindrome(x int) bool {
	// 负数或者各位数为 0 时，不是回文数
	if x < 0 || x != 0 && x%10 == 0 {
		return false
	}

	// 当原始数字小于或等于反转后的数字时，表示已经反转了一半的数字
	reverseNum := 0
	for reverseNum < x {
		reverseNum = reverseNum*10 + x%10
		x /= 10
	}

	// 偶数位数和奇数位数
	return reverseNum == x || reverseNum/10 == x

}
